ipfsconn/ipfshttp/ipfshttp.go

```diff
@@ -450,6 +459,97 @@ func checkResponse(path string, code int, body []byte) error {
 	return fmt.Errorf("IPFS-post '%s' unsuccessful: %d: %s", path, code, body)
 }
 
+// postCtxStreaming makes a POST request against
+// the ipfs daemon, reads the body of the response line by line
+// at evey millisecond. It cancels the context if timeout duration
+// has passed since last new data on the buffer.
+func (ipfs *Connector) postCtxStreaming(ctx context.Context, path string, timeout time.Duration) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	elapsed := time.Now()
+
+	go func() {
+		for {
+			select {
+			case <-timer.C:
+				{
+					if time.Since(elapsed) >= timeout {
+						cancel()
+						break
+					}
+					timer.Reset(timeout - time.Since(elapsed))
+				}
+			}
+		}
+	}()
+
+	res, err := ipfs.doPostCtx(ctx, ipfs.client, ipfs.apiURL(), path, "", nil)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+	reader := bufio.NewReader(res.Body)
+
+	if strings.HasPrefix(path, "pin/add") {
+		var pins ipfsPinsResp
+		var progress int
+		for {
```

Comment:
make a json `Decoder ` with the response reader and run `for { ... dec.Decode(&respObject) ... }` (or with `for dec.More() { ... }`) there is an example in the json module docs.